doc/dev/kdb-record.md

```diff
@@ -0,0 +1,110 @@
+# `kdb record` Command
+
+The `kdb record` tooling enables the user to record configuration sessions. It also allows exporting and importing the configuration changes.
+
+Recording sessions are stored on a per-user basis. This implies that a recording session can only track changes made by a single user.
```

Comment:
Maybe we could mount `user:/elektra/record` with a (custom) resolver that only uses the `$HOME` to detect the user directory?

Reply:
Such a resolver would fail, if `$HOME` isn't set. It also kinda defeats the point of `user:/`.

I'm not sure what the correct solution is but, storing information about `system:/` and/or `spec:/` keys in the `user:/` namespace was weird from the beginning. You could store recordings in `system:/elektra/record`, but then you get into permissions issues. With a single mountpoint you either always need `sudo`, or anybody can write into any session. So for proper permissions, you'd need to create a new mountpoint for every session, then the file could be owned by the user that started the session. But creating mountpoints again requires `sudo`.

Maybe we should store the recordings outside the KDB entirely. I'm think of something like this:
1. Every session is stored in a single file. The path of the file serves as the unique ID (no session_name needed)
2. `kdb record start [session_file]` takes the path to the session file, if none is given it creates one in the current directory. If the given file exists the session will be resumed, otherwise a new file is created.
3. The session file uses a line-based format (something custom or, e.g. [JSON lines](https://jsonlines.org/)). That way adding a new entry to a session is easy, just append a line. It also avoids the need for sequence numbers or unique timestamps, the ordering is given by the order of the lines.
4. The active session is defined by `/elektra/record/session` (normal cascading lookup). `kdb record start` writes to `user:/elektra/record/session` by default, but to cover the `sudo` case you can use `sudo kdb record start --system` with writes to `system:/elektra/record/session` which starts a system-wide session. Alternatively, you could call `kdb record start` and `sudo kdb record start` with the same file to start the session for both the current and the root user, as along as both users have access to the file this should work. System-wide (or even root) sessions could of course be problematic on multi-user systems, because a different user could modify the KDB (possibly with `sudo`) simultaneously which would interfere with the recording. So as a final option, I would suggest that the `ELEKTRA_RECORD_SESSION` environment variable is also read (*) and takes precedence. Then you can simply set the env-var (has to be done manually) and as long as you stay in the same shell `sudo` changes will be recorded correctly.

By storing recordings outside the KDB you also don't need to make sure you exclude the `user:/elektra/record` keys from the recording.

Item 4 from above, could be applied to your current setup as well. But then you still need to solve the problem of where to store the recording.

---

(*) We could use `gopts` for that, but IMO that would be overkill here. Directly reading `ELEKTRA_RECORD_SESSION` and putting it into `proc:/elektra/record/session` (ensures precedence) either during `kdbOpen` or during `kdbGet`, is much simpler and also does the job.

---

If you really want to store everything in the KDB, there is another option. It's not pretty, but it would work. Let the env-var `ELEKTRA_RECORD_USER` define the user for which the changes should be recorded, if the root user causes a change. In other words, if you set `ELEKTRA_RECORD_USER=kodebach` before calling `sudo kdb set system:/foo 123` then the change will be recorded in the active session of user `kodebach` instead of the root user's session. I say this is not pretty, because it requires special case handling for when the root users changes something (not so hard), and writing to another user's `user:/` namespace (quite a bit harder).

Reply:
Yes it looks like storing the session externally is the best option. 

Not sure though that JSON is the best format for this, as storing binary data is pretty cumbersome. I should probably evaluate some other formats like BSON or protobuf as well.

Question is whether @markus2330 approves of adding a new dependency to the `kdb` utility. We could make it optional at build-time though and only include the `kdb record` commands only when it is included.

Reply:
> Not sure though that JSON is the best format for this, as storing binary data is pretty cumbersome.

That's true.

> I should probably evaluate some other formats like BSON or protobuf as well.

To make the whole "just append to a file" idea work, you need a format that has streaming support. If you need to wrap the whole file into something like a JSON array, then "just append to a file" doesn't work.

If you want to look at protobuf, maybe also check out [flatbuffers](https://google.github.io/flatbuffers/). It's similar but designed to require minimal parsing/unpacking. The format can even be `mmap`ed.

> adding a new dependency to the `kdb` utility.

I see a few options:
1. Use a sufficiently simple format that we can implement the parsing ourselves.
2. Use a library that can be linked statically (e.g. I think for flatbuffers this can be done)
3. Make `kdb record` optional as you suggest.

Reply:
@atmaxinger Can you please summarize pros/cons of external serialization in a decision?

I think it is very elegant to have everything within KDB:

1. we can use all our plugins as storage format
2. users can read/manipulate the recordings with their own tooling easily

But if there are compelling reasons against it (sudo alone imho is not, this could be fixed by having "system" sessions that are also accessible by a user, e.g. because it is mounted to /tmp), we'll look for alternatives. Currently the only feature in Elektra that has external data is the cache.

Reply:
> by having "system" sessions that are also accessible by a user

That doesn't work in a true multi-user environment. The system session would be active for all users and that would be nightmare. The environment variable trick, was the only solution I could come up with for this problem.

> I think it is very elegant to have everything within KDB [...]

I guarantee you we will again run into the limitations of the `KeySet` structure. For one you are limit to those formats that can store a `KeySet` 1:1 with all metadata, non-leaf values, binary data, arbitrary names etc.

Some other reasons:
- It's kind of an abuse of the KDB. The KDB is meant to store configuration data. This is quite clearly **not** configuration data.
- Users can't choose the storage plugin anyway, since we store data below `/elektra` where adding mountpoints is prohibited.
- I don't think reading/manipulating the recordings with existing tooling would really be easy, if we force the recording into a `KeySet` format. The format with the key names as basenames needs an extra parsing step to actually know what's going on. You need to first parse the storage file as normal, but then you need to handle the key names as basenames before you actually have the parsed the data fully.
- Using something like "JSON lines" would actually be incredibly easy to handle with existing tools. Just using the `jq` tool already gets you very far. The problem is of course binary data, but Base64 is always an option, or we could support a text-based and a binary format.

---

One thing we could do is a mixed solution. We use a non-KDB wrapper format for the log entries, but each entry contains an `old` and a `new` field which are encoded by the plugin specified in `format`. This would also take care of the issue of changes to multiple keys, since we are recording a full `KeySet` for `old` and `new`.

Binary values will still be a problem, if the wrapper format is text-based. But again we can use Base64, or we use a binary wrapper format. The wrapper format could even be implemented in plugins. Although those would have to be separate from the normal storage plugins.

Here are some examples with different wrapper formats:

JSON Lines
```jsonl
{ "parent": "user:/foo", "format": "dump", "old": "kdbOpen 2\n$end\n", "new": "kdbOpen 2\n$key string 0 3\n\n123\n$end\n"}
{ "parent": "user:/foo", "format": "dump", "old": "kdbOpen 2\n$key string 0 3\n\n123\n$end\n", "new": "kdbOpen 2\n$key string 0 4\n\n1234\n$end\n"}
```

YAML
```yaml
parent: "user:/foo"
format: dump
old: |
  kdbOpen 2
  $end
new: |
  kdbOpen 2
  $key string 0 3

  123
  $end
---
parent: "user:/foo"
format: dump
old: |
  kdbOpen 2
  $key string 0 3
  
  123
  $end
new: |
  kdbOpen 2
  $key string 0 4
  
  1234
  $end
```

Reply:
To get good results in a reasonable time we must take one step after the other:

1. Requirements+Description of user-facing parts
2. Major decisions (structure of log format, internal/external, global plugins, tooling, etc) and architecture
3. Details like which extra libraries, how to implement etc.

Currently we are at step 1, slowly progressing to step 2. Now is not yet the time to discuss syntax, formats etc. (Step 3) They easily bring us astray, as happened in this discussion. Let us focus again on requirements.

The question is which of the following statements need to be captured by the same session:

1. `kdb set dir:/some/key`
2. `kdb set user:/some/key`
3. `sudo -u other_user kdb set user:/some/key`
4. `sudo kdb set system:/some/key`
5. `kdb set user:/some/key` executed by another user

also answering the question:

Do we want the record tooling to work different to Elektra, which knows system, user and dir scopes?

I would answer the second question with no and would not expect 3-5 to end up in a user-session. I would expect all of the calls to end up in a system-session.

From your earlier statements, related to these requirements:

> But creating mountpoints again requires sudo.

Not if we finally implement #1074. Maybe it is related to how recording should look like, as also for recording there might be use cases to have recordings for a system, for a user and for a specific directory.

> Users can't choose the storage plugin anyway, since we store data below /elektra where adding mountpoints is prohibited.

I don't see a fundamental problem, we could easily allow well-defined mountpoints within /elektra, in specific for recording.

Reply:
I see this is quickly becoming one of those discussions with 1000 questions and no answers, where all arguments go straight past each other. I will address some of your concerns, but my main response is at the bottom and in [a separate comment outside this thread.](https://github.com/ElektraInitiative/libelektra/pull/4371#issuecomment-1152903574).

---

You did not address my concerns, about system sessions in multi-user environments. Let's say Alice needs to record a sequence that modifies both `user:/` end `system:/` keys. To achieve her goal, Alice starts a session `alice1` in both user and in system mode.

```sh
kdb record start alice1           # creates and activates the session for current user
sudo kdb record start alice1  # creates and activates the session system-wide
```

Now all changes that Alice makes are recorded in the user/system session. But what happens if at the same time, Bob also uses the machine and makes some changes to `system:/` keys. These would be recorded in the `alice1` session, since the session is active system-wide. Even worse, what if Bob also wants to start a system session

```sh
sudo kdb record start bob1
```

Now some of Alice's changes will be in `alice1` (maybe with some stuff from Bob), but the rest will go into `bob1`.

Unless I misunderstood what you mean by "system sessions", this cannot work like this. There is only one `system:/elektra/record/config/session/current/name` key for the whole system and we can't use a key in `user:/` to define the active system session, because that once again won't be accessible to the root user.

This problem totally negates your question "which of the following statements need to be captured by the same session". If the whole concept doesn't work, there is no need to decide what is recorded in which session.

I think this "system session" concept could only work, if we use an environment variable to define the active session instead. That wouldn't be shared between users. But you didn't even react to this suggestion at all...

---

To address some other things:

>> But creating mountpoints again requires sudo.
>
> Not if we finally implement https://github.com/ElektraInitiative/libelektra/issues/1074.

IIRC the decision there was that `user:/elektra/mountpoints` can only defined `user:/` or `dir:/` mountpoints, but not `system:/` ones. So that's not gonna help.

> I don't see a fundamental problem, we could easily allow well-defined mountpoints within /elektra

True, that issue might not actually be a problem.

---

With all of that out of the way, I agree with you that we need to go step-by-step here. However, IMO we should at least partially consider everything from the beginning. It is useless, if we come up with a perfect description of requirements, only to discover that those requirements cannot be put into practice.